src/server/game/Entities/Object/Object.cpp

```diff
@@ -974,6 +974,18 @@ m_currMap(nullptr), m_InstanceId(0), m_phaseMask(PHASEMASK_NORMAL), m_notifyflag
 {
     m_serverSideVisibility.SetValue(SERVERSIDE_VISIBILITY_GHOST, GHOST_VISIBILITY_ALIVE | GHOST_VISIBILITY_GHOST);
     m_serverSideVisibilityDetect.SetValue(SERVERSIDE_VISIBILITY_GHOST, GHOST_VISIBILITY_ALIVE);
+
+    m_heartBeatTimer.Reset(SECONDS_PER_HEARTBEAT);
+}
+
+void WorldObject::Update(uint32 time_diff)
+{
+    m_heartBeatTimer.Update(time_diff);
+    while (m_heartBeatTimer.Passed())
```

Comment:
while?

Reply:
@ccrs Yep. If since the last update we have missed one or more "heartbeats", we must handle them in the current update.

Reply:
interesting, can you elaborate on the reasoning for this? I cant seem to find it myself

Reply:
@ccrs Heartbeat should fire every 5 seconds. In addition, we need to make sure that if the server has been lagging for the last 10 seconds, Heartbeat will be called twice. We should not miss heartbeats due to lags, because then some heartbeat-periodic aura can proc fewer times than was intended. Or we'll miss something else important.

Timer of normal periodic auras works like this. You can see here:  https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Spells/Auras/SpellAuraEffects.cpp#L820

Reply:
Imagine that we're about to execute the tick, but the server freezes for 7 seconds. Now we're two ticks behind on heartbeat, the first one was the one we were about to execute, and the second one was meant to come 5 secs later, but the server was frozen at that time. With the "if" approach we have:
```
// timer is at -7s
if (timer <= 0)
{
    tick();
    timer = 5s;
}
```
"if" block gets executed once, timer is now set to 5s. We ticked only once, even tho we should've ticked twice. The timer is now at 5s, instead of 3s, as it should've been after 7s of lag.
Now the "while" approach:
```
// timer is at -7s;
while (timer <= 0)
{
    tick();
    timer += 5s;
}
```
"while" block get executed the first time, timer is now at -2s, "while" block gets executed the second time, timer is now at 3s. We ticked twice as intended. Yes, these ticks are not spaced out in time as intended, but at least their NUMBER is consistent, and the state of timer after the lag is preserved so it doesn't get out of sync.

This is the idea behind using "while" for timer checks. Is it needed here? Not really. There are a lot of other places where it would be more useful, for example in periodic auras. There is just one problem with this approach: if the lag was CAUSED by the tick - it will only make things spiral out of control even further and snowball the lag to the point where it would just freeze the server completely. Using "if" sort of prevents this from escalating too much, but at the cost of consistency.

Reply:
I see thanks, for the explaination.

 I did miss the bellow statement
```c++
m_heartBeatTimer.Reset(m_heartBeatTimer.GetExpiry() + SECONDS_PER_HEARTBEAT);
```

neat, I'm now considering if this approach should be also applied to movement generators respective timers...